pyccel/ast/headers.py

```diff
@@ -212,7 +234,31 @@ def build_argument(var_name, dc):
                 dtypes += [[i]]
             else:
                 raise TypeError('element must be of type UnionType or dict')
-        for args_ in product(*dtypes):
+
+        #TODO: handle the case of functions arguments
+        arg_codes = []
+
+        #generate possible combinations from uniontypes and iterate through them
+        for iterx in product(*dtypes):
+            iterx = list(iterx)
+            old_values = {}
+            #For each template type, replace only one in each combination, by
+            #a list of its args
+            for i, j in enumerate(iterx):
+                if 'datatype' in j and j['datatype'] in templates and not j['datatype'] in old_values:
+                    iterx[i] = templates[j['datatype']].args
+                    old_values[j['datatype']] = i
+                else:
+                    iterx[i] = [j]
+            #generate combinations using product, and replace remaining
+            #templates with its appropriates values
+            for itery in product(*iterx):
+                itery = list(itery)
+                for j, i in enumerate(itery):
+                    if 'datatype' in i and i['datatype'] in templates:
+                        itery[j] = itery[old_values[i['datatype']]].copy()
+                arg_codes.append(itery)
```

Comment:
I find this code difficult to understand, probably because of the choice of variables' names. I recommend using `i` and `j` for the integer indices, but not for the actual elements of an iterable container.

It also seems to me that the templates could be processed one by one, until they are all used. I am giving here some pseudo-code to explain what I mean:
```python
# First, process unions
...

# Next, process templates
arg_codes = (*dtypes)
for t in templates:
    arg_codes = (replace_template(c, t, a) for c in arg_codes for a in t.args)
```
Where function `replace_template` creates a new function signature by replacing a certain template `T` with one of its possible values. Please let me know if I am missing something here!

Reply:
I think this is a better way to do it, I will try to improve this code section